texar/tf/utils/dynamic_decode.py

```diff
@@ -0,0 +1,314 @@
+from tensorflow.python.framework import tensor_shape
+from tensorflow.python.util import nest
+
+
+__all__ = [
+    "dynamic_decode"
+]
+
+
+def dynamic_decode(decoder, output_time_major=False, impute_finished=False,
+                   maximum_iterations=None, parallel_iterations=32,
+                   swap_memory=False, scope=None, **kwargs):
+  """Perform dynamic decoding with `decoder`.
+  Calls initialize() once and step() repeatedly on the Decoder object.
+  Args:
+    decoder: A `Decoder` instance.
+    output_time_major: Python boolean.  Default: `False` (batch major).  If
+      `True`, outputs are returned as time major tensors (this mode is faster).
+      Otherwise, outputs are returned as batch major tensors (this adds extra
+      time to the computation).
+    impute_finished: Python boolean.  If `True`, then states for batch
+      entries which are marked as finished get copied through and the
+      corresponding outputs get zeroed out.  This causes some slowdown at
+      each time step, but ensures that the final state and outputs have
+      the correct values and that backprop ignores time steps that were
+      marked as finished.
+    maximum_iterations: `int32` scalar, maximum allowed number of decoding
+       steps.  Default is `None` (decode until the decoder is fully done).
+    parallel_iterations: Argument passed to `tf.while_loop`.
+    swap_memory: Argument passed to `tf.while_loop`.
+    scope: Optional variable scope to use.
+    **kwargs: dict, other keyword arguments for dynamic_decode. It might contain
+      arguments for `BaseDecoder` to initialize, which takes all tensor inputs
+      during call().
+  Returns:
+    `(final_outputs, final_state, final_sequence_lengths)`.
+  Raises:
+    TypeError: if `decoder` is not an instance of `Decoder`.
+    ValueError: if `maximum_iterations` is provided but is not a scalar.
+  """
+  if not isinstance(decoder, (TFDecoder)):
+        raise TypeError("Expected decoder to be type Decoder, but saw: %s" %
+                        type(decoder))
+
+  with tf.variable_scope(scope, "decoder") as varscope:
+    # Determine context types.
+
+    if maximum_iterations is not None:
+        maximum_iterations = tf.convert_to_tensor(
+            maximum_iterations, dtype=tf.int32, name="maximum_iterations")
+        if maximum_iterations.get_shape().ndims != 0:
+            raise ValueError("maximum_iterations must be a scalar")
+
+    if isinstance(decoder, TFDecoder):
+        initial_finished, initial_inputs, initial_state = decoder.initialize()
+    else:
+        # For BaseDecoder that takes tensor inputs during call.
+        decoder_init_input = kwargs.pop("decoder_init_input", None)
+        decoder_init_kwargs = kwargs.pop("decoder_init_kwargs", {})
+        initial_finished, initial_inputs, initial_state = decoder.initialize(
+            decoder_init_input, **decoder_init_kwargs)
+
+    def _concat(prefix, suffix, static=False):
```

Comment:
Why do we define this function? I cannot find it in `tensorflow`'s `dynamic_decode` function. @TomNong